### PR TITLE
Smart Contract Upgradability with ZeppelinOS

### PR DESCRIPTION
**Pull request by mestorlx**
## Description
This is a POC of how `zos` can be used to manage upgradability on keeper-contracts. All contracts need to be updated and migrations should be replaced with a deployment script.  The test can be run with `npx truffle test test/upgradability/DIDRegistry.Test.js`.  
To run the test locally only `DIDRegistry.sol`, `DIDRegistryUpgrades.sol` and `MultiSigWallet.sol` should be compiled since `zos` use a different version of `oppenzeppelin`. 
Migrations folder should be empited since they are not directly supported by `zos`

## Is this PR related with an open issue?
close oceanprotocol/ocean-bounties#09

## Types of changes
Breaking change (fix or feature that would cause existing functionality to change)
 
## Checklist:

- [x] Follows the code style of this project.
- [x] Tests Cover Changes
- [x] Documentation

#### Funny gif

Put a link of a funny gif inside the parenthesis-->

### CODE REVIEW

**Comment by mestorlx** [comment]
hi @gongf05 thanks for the review. A few notes.
> Thanks for the update! I found some work had been done as:

Another change that is required is to remove the migrations (including Migrations.sol) since all setup of the contracts is done with the script `deployContracts.sh` and `zos` does not support migrations directly. 

> It would be better to add following changes:
> 
> * add README file of step-by-step instruction including how to make changes, how to deploy, how to upgrade, and how to test.
> * for each step, please include the invoked command, its purpose, expected output, and screenshot of real output.
Will add
> * add more comprehensive testing: change storage only, change logic only, change interface, etc. Please compare the expected output vs. real testing output.
> * make the similar changes to all smart contracts in Keeper contracts.
> * fix the build error in Travis CI.

I was planning on completing these once we agree that what was done for `DIDRegistry` its the desired approach. 

> * suggest pros and cons of zeppelinOS.

will add

**Comment by mestorlx** [comment]
Hi @gongf05 and @H34D 
I have made some progress, please see below. Will be adding the missing contracts next, after that will fix all travis CI errors.

Check list before merge:

- [x] Add README file of step-by-step instruction including how to make changes, how to deploy, how to upgrade, and how to test.
- [x] Add more comprehensive testing: change storage only, change logic only, change interface, etc.
- [x] Suggest pros and cons of zeppelinOS.
- [ ] Make similar changes to all smart contracts in Keeper contracts.
- [ ] Fix the build error in Travis CI.
- [ ] Follows the code style of this project.

ZeppelinOS Advantages:

- Allows to upgrade functionality.
- Allows to add storage variables.
- Allows to use deployed contracts reducing audits and gas cost.
- The proxy pattern requires few changes to current contracts.
- `zos` will perform some checks to validate upgrades.

ZeppelinOS Disadvantage:

- Can not remove storage variables and certain limitations in modifying [declared variables](https://github.com/jackandtheblockstalk/upgradeable-proxy#332-you-cant-thumbsdown).
- Manually keep track of the initialize function.
- Constructors can not be used.
- Admin has full control over upgradability although this can be mitigated granting admin rights to a trustless contract.
- Vulnerable to attack vectors in deployed standard libraries.
- Not particularly an issue with `zos` but with upgradability itself: Using a proxy to change the smart contract logic does not violates the notion of immutability?

List of Changes:

- Constructors are replaced by initialization functions
- Migrations are replaced by deployment script
- Artifacts folder will be replaced by zos.network.json where all proxies will be stored.
- Upgrade openzeppelin-solidity to openzeppelin-eth.
- Add `ContractUpgrades.sol` with upgraded version of the contracts for testing.

**Comment by mestorlx** [comment]
Quick question not related to this PR but I just saw that `DIDregistry` is `Ownable` but no method is protected by `onlyOwner` and `OceanToken` import `Ownable` however it is not being used. Any reason for this @gongf05 ?

**Comment by H34D** [comment]
> Thanks for the update! I found some work had been done as:
> 
> * change openzeppelin-solidity to be openzeppelin-eth to reuse the EVM package.
> * change constructor function of DIDRegistry contract to be "initialize" to make it to be upgradeable contract.
> * modify DIDRegistry contract to be "DIDRegistryUpgrades.sol" (DIDRegistry V2.0 contract for testing)
> * add multiSig wallet contract for upgradeability governance as instructed in tutorial: https://github.com/zeppelinos/labs/tree/master/governance-with-multisig
> * add deployment script
> * add testing file "DIDRegistry.Test.js" to walk through the upgrade process.
> 
> It would be better to add following changes:
> 
> * add README file of step-by-step instruction including how to make changes, how to deploy, how to upgrade, and how to test.
> * for each step, please include the invoked command, its purpose, expected output, and screenshot of real output.
> * add more comprehensive testing: change storage only, change logic only, change interface, etc. Please compare the expected output vs. real testing output.
> * suggest pros and cons of zeppelinOS.
> * make the similar changes to all smart contracts in Keeper contracts.
> * fix the build error in Travis CI.

is this resolved already? @gongf05

**Comment by gongf05** [comment]
@mestorlx Thanks for great work! It resolved most of concerns.

Some questions before we wrap up this work:
1. contract "SLA/ServiceAgreement.sol" is not using "initializer" so it will not be able to be upgraded. any reason for that?

2. the doc "upgrades.md" mentioned "./script/setupWallet.sh", which cannot be found. 

3. the command "$ ./scripts/deployContracts.sh " fails. 
I need to add "OWNER=accounts[0]" and remove " --skip-compile" in "npx zos push --skip-compile  -v" to make it pass. Is that correct change? Below is the failure screenshot.

keeper-contracts_ _node_ _deploycontracts_sh_ _174x66

4. The testing is somewhat not stable, whcih may be the issue of zeppelinOS. the same test can fail first and pass in the second run. 

5. could you provide any suggestions regarding zeppelinOS? any pros & cons?

Thanks,

File: contracts/OceanMarket.sol [hunk 1]
```diff
@@ -1,7 +1,8 @@
 pragma solidity 0.4.25;
 
-import 'openzeppelin-solidity/contracts/math/SafeMath.sol';
-import 'openzeppelin-solidity/contracts/ownership/Ownable.sol';
+import 'openzeppelin-eth/contracts/math/SafeMath.sol';
```
Comment (H34D) [inline contracts/OceanMarket.sol]: Please switch back to `openzeppelin-solidity` `openzeppelin-eth` is not longer maintained. We need to migrate to solidity `v0.5.0` soon that is only supported with `openzeppelin-solidity`

File: contracts/OceanMarket.sol [hunk 2]
```diff
@@ -1,7 +1,8 @@
...
+import 'openzeppelin-eth/contracts/ownership/Ownable.sol';
```
Comment (H34D) [inline contracts/OceanMarket.sol]: Please switch back to `openzeppelin-solidity` `openzeppelin-eth` is not longer maintained. We need to migrate to solidity `v0.5.0` soon that is only supported with `openzeppelin-solidity`

File: contracts/SLA/ComputeConditions.sol [hunk 3]
```diff
@@ -1,14 +1,15 @@
 pragma solidity ^0.4.25;
 
-import 'openzeppelin-solidity/contracts/cryptography/ECDSA.sol';
+import 'zos-lib/contracts/Initializable.sol';
+import 'openzeppelin-eth/contracts/cryptography/ECDSA.sol';
```
Comment (H34D) [inline contracts/SLA/ComputeConditions.sol]: Please switch back to `openzeppelin-solidity` `openzeppelin-eth` is not longer maintained. We need to migrate to solidity `v0.5.0` soon that is only supported with `openzeppelin-solidity`

File: contracts/SLA/FitchainConditions.sol [hunk 4]
```diff
@@ -1,6 +1,7 @@
 pragma solidity ^0.4.25;
 
-import 'openzeppelin-solidity/contracts/math/SafeMath.sol';
+import 'zos-lib/contracts/Initializable.sol';
+import 'openzeppelin-eth/contracts/math/SafeMath.sol';
```
Comment (H34D) [inline contracts/SLA/FitchainConditions.sol]: Please switch back to `openzeppelin-solidity` `openzeppelin-eth` is not longer maintained. We need to migrate to solidity `v0.5.0` soon that is only supported with `openzeppelin-solidity`

File: contracts/SLA/PaymentConditions.sol [hunk 5]
```diff
@@ -1,14 +1,15 @@
 pragma solidity 0.4.25;
 
-import 'openzeppelin-solidity/contracts/token/ERC20/ERC20.sol';
+import 'openzeppelin-eth/contracts/token/ERC20/ERC20.sol';
```
Comment (H34D) [inline contracts/SLA/PaymentConditions.sol]: Please switch back to `openzeppelin-solidity` `openzeppelin-eth` is not longer maintained. We need to migrate to solidity `v0.5.0` soon that is only supported with `openzeppelin-solidity`

File: contracts/SLA/ServiceAgreement.sol [hunk 6]
```diff
@@ -1,6 +1,6 @@
 pragma solidity 0.4.25;
 
-import 'openzeppelin-solidity/contracts/cryptography/ECDSA.sol';
+import 'openzeppelin-eth/contracts/cryptography/ECDSA.sol';
```
Comment (H34D) [inline contracts/SLA/ServiceAgreement.sol]: Please switch back to `openzeppelin-solidity` `openzeppelin-eth` is not longer maintained. We need to migrate to solidity `v0.5.0` soon that is only supported with `openzeppelin-solidity`

File: contracts/didregistry/DIDRegistry.sol [hunk 7]
```diff
@@ -1,8 +1,9 @@
 pragma solidity 0.4.25;
 
-import 'openzeppelin-solidity/contracts/ownership/Ownable.sol';
+import 'zos-lib/contracts/Initializable.sol';
+import 'openzeppelin-eth/contracts/ownership/Ownable.sol';
```
Comment (H34D) [inline contracts/didregistry/DIDRegistry.sol]: Please switch back to `openzeppelin-solidity` `openzeppelin-eth` is not longer maintained. We need to migrate to solidity `v0.5.0` soon that is only supported with `openzeppelin-solidity`

File: contracts/testContracts/DIDRegistryUpgrades.sol [hunk 8]
```diff
@@ -0,0 +1,250 @@
+/* solium-disable */
+pragma solidity 0.4.25;
+
+// Contain upgraded version of the contracts for test
+import 'zos-lib/contracts/Initializable.sol';
+import 'openzeppelin-eth/contracts/ownership/Ownable.sol';
```
Comment (H34D) [inline contracts/testContracts/DIDRegistryUpgrades.sol]: Please switch back to `openzeppelin-solidity` `openzeppelin-eth` is not longer maintained. We need to migrate to solidity `v0.5.0` soon that is only supported with `openzeppelin-solidity`

File: contracts/testContracts/OceanTokenUpgrades.sol [hunk 9]
```diff
@@ -0,0 +1,357 @@
+/* solium-disable */
+pragma solidity 0.4.25;
+
+import 'openzeppelin-eth/contracts/token/ERC20/ERC20.sol';
```
Comment (H34D) [inline contracts/testContracts/OceanTokenUpgrades.sol]: Please switch back to `openzeppelin-solidity` `openzeppelin-eth` is not longer maintained. We need to migrate to solidity `v0.5.0` soon that is only supported with `openzeppelin-solidity`

File: contracts/token/OceanToken.sol [hunk 10]
```diff
@@ -1,37 +1,43 @@
 pragma solidity 0.4.25;
 
-import 'openzeppelin-solidity/contracts/token/ERC20/ERC20.sol';
-import 'openzeppelin-solidity/contracts/ownership/Ownable.sol';
+import 'openzeppelin-eth/contracts/token/ERC20/ERC20.sol';
```
Comment (H34D) [inline contracts/token/OceanToken.sol]: Please switch back to `openzeppelin-solidity` `openzeppelin-eth` is not longer maintained. We need to migrate to solidity `v0.5.0` soon that is only supported with `openzeppelin-solidity`

File: package.json [hunk 11]
```diff
@@ -53,10 +53,13 @@
     "ethjs": "^0.4.0",
     "ethjs-provider-http": "^0.1.6",
     "ethlint": "^1.2.1",
-    "openzeppelin-solidity": "^2.0.0",
+    "glob": "^7.1.3",
+    "openzeppelin-eth": "^2.0.2",
```
Comment (H34D) [inline package.json]: Please switch back to `openzeppelin-solidity` `openzeppelin-eth` is not longer maintained. We need to migrate to solidity `v0.5.0` soon that is only supported with `openzeppelin-solidity`

File: package.json [hunk 12]
```diff
@@ -9,15 +9,15 @@
...
     "migrate:nile": "npm run migrate -- --network nile",
     "migrate:kovan": "npm run migrate -- --network kovan",
     "migrate:rinkeby": "npm run migrate -- --network rinkeby",
     "migrate:mainnet": "npm run migrate ----network mainnet",
-    "test": "truffle test",
-    "test:cover": "solidity-coverage",
-    "report-coverage": "cat ./coverage/lcov.info | codacy-coverage --token b95545c0da4b45848cc5eef363c2711e --prefix ./",
+    "test": "if [ -z ${NETWORK} ]; then export NETWORK='development'; fi && if [ ! -f wallet.json ]; then truffle exec scripts/setupWallet.js --network $NETWORK; fi && truffle test",
+    "test:cover": "export NETWORK='coverage' && solidity-coverage",
+    "report-coverage": "cat ./coverage/lcov.info | codacy-coverage --token b95545c0da4b45848cc5eef363c2711e",
```
Comment (H34D) [inline package.json]: please add `--prefix ./` back again

Comment (mestorlx) [reply package.json]: Done!

**Pull request merged**